haystack_experimental/utils/jinja_chat_extension.py

```diff
@@ -0,0 +1,203 @@
+# SPDX-FileCopyrightText: 2022-present deepset GmbH <info@deepset.ai>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import json
+from typing import Any, Callable, List, Optional, Union, get_args
+
+from haystack import logging
+from jinja2 import TemplateSyntaxError, nodes
+from jinja2.ext import Extension
+
+from haystack_experimental.dataclasses.chat_message import (
+    ChatMessage,
+    ChatMessageContentT,
+    ChatRole,
+    TextContent,
+    _deserialize_content_part,
+    _serialize_content_part,
+)
+
+logger = logging.getLogger(__name__)
+
+START_TAG = "<haystack_content_part>"
+END_TAG = "</haystack_content_part>"
+
+class ChatMessageExtension(Extension):
+    """
+    A Jinja2 extension for creating structured chat messages with mixed content types.
+
+    This extension provides a custom {% message %} tag that allows creating chat messages
+    with different attributes (role, name, meta) and mixed content types (text, images, etc.).
+
+    Inspired by [Banks](https://github.com/masci/banks).
+
+    Example:
+    {% message role="system" %}
+    You are a helpful assistant. You like to talk with {{user_name}}.
+    {% endmessage %}
+
+    {% message role="user" %}
+    Hello! I am {{user_name}}. Please describe the images.
+    {% for image in images %}
+    {{ image | for_template }}
+    {% endfor %}
+    {% endmessage %}
+
+    ### How it works
+    1. The {% message %} tag is used to define a chat message.
+    2. The message can contain text and other structured content parts.
+    3. To include a structured content part in the message, the `| for_template` filter is used.
+       The filter serializes the content part into a JSON string and wraps it in a `<haystack_content_part>` tag.
```

Comment:
Does this mean manual insertion of `<haystack_content_part>` isn't expected of users to do? Or would it be a good option for advanced users?

Reply:
While this is technically feasible, I would say that if we design our abstraction well, I hope that there is no need for users to know the underlying mechanism and interfere with it. It would just be an implementation detail that we can improve/alter in the future (without worrying about anyone being impacted).